### ISSUE TITLE
DOC: numpy 2: deprecated type cast behaviour change not mentioned in release notes

### ISSUE THREAD

**Issue by illume** [numpy/numpy#26596]
### Describe the issue:

Before sending 255 to numpy array in8 would be converted to -1. Now it raises an error.

This is not mentioned in the migration guide, and the linter does not catch it.

numpy 1.x
```py
import
numpy.array(numpy.array([[0, -128], [-100, 100], [37, -80], [255, 0]], numpy.int8))
array([[   0, -128],
       [-100,  100],
       [  37,  -80],
       [  -1,    0]], dtype=int8)
```

numpy 2.x
```py
import
numpy.array(numpy.array([[0, -128], [-100, 100], [37, -80], [255, 0]], numpy.int8))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OverflowError: Python integer 255 out of bounds for int8
```

This means all code using arrays needs proper testing/review. Because without it, it's easy to have tests for numbers within range, but when the code encounters these issues an error will be raises. Int 8 is not so bad, but 64bit / and such will probably not be so well tested. So I expect users code will raise errors here in production.

It's not mentioned in the migration guide, and the linter does not catch it.  Additionally I didn't see a type check change... but maybe a type checker can catch this too (note I didn't test this well... so maybe some type checker already catches this).

### Reproduce the code example:

```python
numpy 1.x

import
numpy.array(numpy.array([[0, -128], [-100, 100], [37, -80], [255, 0]], numpy.int8))
array([[   0, -128],
       [-100,  100],
       [  37,  -80],
       [  -1,    0]], dtype=int8)
```

numpy 2.x
```py
import
numpy.array(numpy.array([[0, -128], [-100, 100], [37, -80], [255, 0]], numpy.int8))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OverflowError: Python integer 255 out of bounds for int8
```

### Python and NumPy Versions:

python 3.12, numpy 2.0.0rc2

**Comment by seberg** [comment]
>  and there was no deprecation notice given for this change.

Just for the record, this was deprecated in 1.24: https://numpy.org/devdocs/release/1.24.0-notes.html#conversion-of-out-of-bound-python-integers and did give DeprecationWarnings:
```
In [2]: numpy.array(numpy.array([[0, -128], [-100, 100], [37, -80], [255, 0]], numpy.int8))
   ...: 
<ipython-input-2-a2ef1427d77a>:1: DeprecationWarning: NumPy will stop allowing conversion of out-of-bound Python integers to integer arrays.  The conversion of 255 to int8 will fail in the future.
For the old behavior, usually:
    np.array(value).astype(dtype)
will give the desired result (the cast overflows).
```

It may be missing release notes because finalizing it was a necessary side-effect of other changes.  So adding docs/reinstating some of that message may be good.

**Comment by illume** [comment]
Thanks @seberg, I've removed the deprecation notice comments from issue description/title. The systems I was testing on had an older numpy (Ubuntu and Debian based).

**Comment by mattip** [comment]
Adding the finalized deprecation to the release notes might be nice. I changed the title to reflect this.

**Comment by seberg** [comment]
Yes, we should.  And if it's not there a brief mention in the 2.0 guide probably makes sense.  It is one of the trickier/noisier one if you didn't already notice the warnings earlier!

**Comment by Searchingdays** [comment]
I tried adding the depreciation in the 2.0 release notes. 

* ``numpy.int8`` will no longer support conversion of out of bounds python integers to integer arrays. For example, conversion of 255 to int8 will not return -1.
  ``np.array(value).astype(dtype)`` will give the desired result.

* ``np.safe_eval`` has been deprecated. ``ast.literal_eval`` should be used instead.

**Comment by mattip** [comment]
Closed in PR #26635
